File: net/JetBrains.Serialization/src/Dataholders/UnknownEncapContentInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using Org.BouncyCastle.Asn1.Cms;

namespace JetBrains.Serialization;

[JsonObject(MemberSerialization.OptIn)]
public class UnknownEncapContentInfo : EncapContentInfo
{
  [JsonProperty("ContentType")] protected override TextualInfo ContentType { get; }
  [JsonProperty("Content")] private IEncodableInfo? _content;

  [JsonConstructor]
  public UnknownEncapContentInfo(TextualInfo contentType, IEncodableInfo? content)
  {
    ContentType = contentType;
    _content = content;
  }

  public UnknownEncapContentInfo(ContentInfo contentInfo)
    : this(TextualInfo.GetInstance(contentInfo.ContentType),
      contentInfo.Content?.ToAsn1Object()?.ToEncodableInfo())
  {
  }

  protected override Asn1Encodable? GetContentPrimitive() => _content?.ToPrimitive();
}

File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/dataholders/UnknownEncapContentInfo.kt
package com.jetbrains.signatureverifier.serialization.dataholders

import com.jetbrains.signatureverifier.serialization.toEncodableInfo
import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.cms.ContentInfo

@Serializable
data class UnknownEncapContentInfo(
  override val contentType: TextualInfo,
  val content: EncodableInfo?
) : EncapContentInfo() {

    constructor(contentInfo: ContentInfo) : this(
      TextualInfo.getInstance(contentInfo.contentType),
        contentInfo.content?.toASN1Primitive()?.toEncodableInfo()
    )

    override fun getContentPrimitive() = content?.toPrimitive()
}